### PR TITLE
VPN-6011: Add a Proxy to allow Containers&Tabs to Selectively Split-Tunnel

### PR DESCRIPTION
**Pull request by strseb**
#tldr; 
This PR adds a sidecar binary, `loophole` - a simple socks5 proxy that runs alongside the client and is avoiding the tunnel. This allows containers or the extension to specify that they do not want this traffic to Use the tunnel.

image

## Description

 - This pr is based on @bakulf's implementation of socks5 https://github.com/mozilla-mobile/mozilla-vpn-client/pull/4022 
 - To the socks5 impl i only removed some duplicate code, added passing the config via args. 
 - I added a Proxycontroller, that can start/stop the proxy if needed and the Webextension class allow Web-Extensions Like FFVPN or MAC to Query the loophole proxy address. This should allow extension to exclude websites on a per-tab/container bases. Also relates to https://github.com/mozilla/multi-account-containers/issues/2259 

I aim to add more tests and maybe add UDP/Auth support, but this pr is already big and the sprint is short.
As is the new binaries are not shipped and the code is disabled if not in prod so we can land this, if we want.

### CODE REVIEW

**Review by lesleyjanenorton — COMMENTED** [review]
This is awesome. Just a few questions/comments so far. I have to step out for an appt and will come back to this right after ✨

File: src/controller.cpp [hunk 1]
```diff
@@ -399,6 +403,12 @@ void Controller::activateInternal(DNSPortPolicy dnsPort,
   // Splittunnel-feature could have been disabled due to a driver conflict.
   if (Feature::get(Feature::Feature_splitTunnel)->isSupported()) {
     exitConfig.m_vpnDisabledApps = settingsHolder->vpnDisabledApps();
+// Add the Proxy to the excluded list, if activateable
+#if defined MZ_PROXY_ENABLED
+    if (vpn->proxyController()->canActivate()) {
+      exitConfig.m_vpnDisabledApps.append(vpn->proxyController()->binaryPath());
```
[the remark this answers is not in the document]
Comment (lesleyjanenorton) [reply src/controller.cpp:409]: Filed for later: https://mozilla-hub.atlassian.net/browse/VPN-6490

[the remark this answers is not in the document]
Comment (strseb) [reply src/controller.cpp:409]: No :) 
given ``` exitConfig.m_vpnDisabledApps = settingsHolder->vpnDisabledApps(); ``` we grab a copy of whatever that list is having and we add the proxy to the copy of the list. in other words if it exists it's always splittunneled.

File: src/socks5proxy/src/socks5client.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+#ifndef SOCKS5CLIENT_H
+#define SOCKS5CLIENT_H
+
+#include <QObject>
+#include <QTcpSocket>
+
+class Socks5;
+
+class Socks5Client final : public QObject {
```
Comment (oskirby) [inline src/socks5proxy/src/socks5client.h:13]: The name `Socks5Client` can be a bit opaque as it could apply to both the browser trying to talk to the proxy, or the proxy being a client to some server out in the internet. I would rename this class to something like `Socks5Connection` or `Socks5Session`.

File: src/socks5proxy/src/socks5.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <QObject>
+#include <QTcpServer>
+
+class Socks5 final : public QObject {
+  Q_OBJECT
+  Q_PROPERTY(uint16_t connections READ connections NOTIFY connectionsChanged);
+
+ public:
+  explicit Socks5(QObject* parent, uint16_t port, QHostAddress listenAddress);
```
Comment (oskirby) [inline src/socks5proxy/src/socks5.h:16]: Minor nit: Qt convention usually puts the `QObject* parent` as the last parameter and gives it a default value of `nullptr` This allows the caller to omit it, in which case it is bound to the lifetime of the `qApp`.

File: src/socks5proxy/src/socks5client.cpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+ServerResponsePacket createServerResponsePacket(uint8_t rep,
+                                                uint16_t port = 0x00u) {
+  return ServerResponsePacket{0x05, rep, 0x00, 0x01, 0x00, port};
+}
+
+}  // namespace
+
+Socks5Client::Socks5Client(Socks5* parent, QTcpSocket* socket, uint16_t port)
+    : QObject(parent), m_parent(parent), m_inSocket(socket), m_socksPort(port) {
+  connect(m_inSocket, &QTcpSocket::disconnected, this, &QObject::deleteLater);
```
Comment (oskirby) [inline src/socks5proxy/src/socks5client.cpp:77]: Just considering the risk of dangling pointers here. Would it make more sense for the `Socks5Client` to be a child of the socket? This would give a hierarchy of: `Socks5` -> `QTcpSocket` -> `Socks5Client`.

Or maybe the `Socks5Client` should take ownership of the provided socket, in which case we would get:
`Socks5` -> `Socks5Client` -> [`m_inSocket`, `m_outSocket`]

The thing we don't want is for some other parent to garbage collect the `QTcpSocket` when the `Socks5Client` is still pointing to it.

**Comment by oskirby** [comment]
This is some great work, thank you!

**Comment by oskirby** [comment]
I am not a huge fan of the name `loophole`. While I get that it is descriptive for what it's doing, it does also kind of imply some shady sneakiness afoot.

File: src/mozillavpn.cpp [hunk 5]
```diff
@@ -1120,6 +1120,11 @@ void MozillaVPN::controllerStateChanged() {
   if (m_updating && m_private->m_controller.state() == Controller::StateOff) {
     update();
   }
+  if (m_private->m_controller.state() == Controller::StateOn) {
+    m_private->m_proxyController.start();
+  } else {
+    m_private->m_proxyController.stop();
```
[the remark this answers is not in the document]
Comment (strseb) [reply src/mozillavpn.cpp:1126]: I think it's a good idea to stop the proxy so all current connections get closed, so that firefox need's to reopen them once we switched.

File: src/mozillavpn.cpp [hunk 6]
```diff
@@ -1120,6 +1120,11 @@ void MozillaVPN::controllerStateChanged() {
   if (m_updating && m_private->m_controller.state() == Controller::StateOff) {
     update();
   }
+  if (m_private->m_controller.state() == Controller::StateOn) {
+    m_private->m_proxyController.start();
```
[the remark this answers is not in the document]
Comment (strseb) [reply src/mozillavpn.cpp:1124]: good call

File: src/socks5proxy/proxycontroller.cpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  if (!mCurrentProcess && std::holds_alternative<Stopped>(m_state.value())) {
+    return;
+  }
+  logger.debug() << "Loophole stopping";
+  // We no longer need signals from this obj
+  m_state = Stopped{};
+  mCurrentProcess->kill();
+}
+
+bool ProxyController::canActivate() {
```
[the remark this answers is not in the document]
Comment (strseb) [reply src/socks5proxy/proxycontroller.cpp:108]: I do not think this is a good check. You can have multiple versions of firefox, and each can have multiple profiles, and it's not required for the profile dir to be in a speficic folder, so this could go wrong, given we'd disable a feature if we fail to detect the extension.

File: src/socks5proxy/bin/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+qt_add_executable(loophole 
+    main.cpp
+)
+
+target_link_libraries(loophole PUBLIC
```
[the remark this answers is not in the document]
Comment (strseb) [reply src/socks5proxy/bin/CMakeLists.txt:9]: Right. Moved it to `/extension` also changed the name :D

File: src/socks5proxy/src/socks5client.h [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+   *
+   * There is no formal typecheck that the underlying
+   * data of <T> is correct.
+   *
+   * @tparam T - Type to read
+   * @param connection - A QIODevice to read from
+   * @return std::optional<T> - The read T if enough bytes were written.
+   */
+  template <typename T>
+  static std::optional<T> readPaket(QIODevice* connection);
```
[the remark this answers is not in the document]
Comment (strseb) [reply src/socks5proxy/src/socks5client.h:62]: yes, accidentally German. :D

**Comment by strseb** [comment]
@oskirby , @lesleyjanenorton - mind doing another round of reviews ? :)

File: extension/socks5proxy/src/socks5connection.cpp [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+      }
+    } break;
+
+    default:
+      Q_ASSERT(false);
+      break;
+  }
+}
+
+qint64 Socks5Connection::proxy(QIODevice* a, QIODevice* b) {
```
Comment (oskirby) [inline extension/socks5proxy/src/socks5connection.cpp:238]: Just scrutinizing this method a little bit more, and I am concerned that it might have some difficulties under heavy traffic loads. In particular when non-blocking IO is used it might be the case that we exhaust the outgoing buffer in `b->write()` and it returns fewer than `val` bytes. This case is not an error, but we need to wait for those bytes to be written (eg: by handling the `bytesWritten` signal) and re-trying the write when there is more space available.

I think this implementation is fine for a proof of concept, but I just wanted to note that it might need a bit more testing and massaging before we ship to users.

**Review by oskirby — APPROVED** [review]

**Comment by oskirby** [comment]
As a general question: what does the CPU usage of this process look like under heavy traffic loading spread across many connections? As written this is single-threaded so there is only one core available to handle the all of the network traffic that goes through this proxy.

**Comment by strseb** [comment]
> As a general question: what does the CPU usage of this process look like under heavy traffic loading spread across many connections? As written this is single-threaded so there is only one core available to handle the all of the network traffic that goes through this proxy.

That is a very good question we certainly could consider using a threadpool to handle this. 
So today i did some boring testing, on my machine (tm) just setting this as my firefox proxy. 

[washpo.webm](https://github.com/user-attachments/assets/acde4d67-5021-43c9-a130-3b6b3767610a)


-> https://www.washingtonpost.com/, without etp+ublock juicy 200+ requests just for the first page. (great design) -  as you can see the CPU impact of the proxy compared to the overall CPU usage (of rendering that page and all the glorious trackers and ads) is pretty looow. 

-> A big thing i.e https://ash-speed.hetzner.com/ and pulling one big file actually put's a dent on the CPU usage, but it's in the <5% range, and it did not make a speed difference. To be fair i am running a ryzen-9 which is fairly powerful.
